src/main/haskell/kore/src/Kore/SMT/SMT.hs

```diff
@@ -0,0 +1,181 @@
+{-|
+Module      : Data.Kore.SMT.SMT
+Description : Basic SMT interface.
+Copyright   : (c) Runtime Verification, 2018
+License     : UIUC/NCSA
+Maintainer  : phillip.harris@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+{-# OPTIONS_GHC -Wno-unused-matches #-}
+
+module Kore.SMT.SMT
+( patternToSMT
+, provePatternIO
+, provePattern
+, provePredicate
+, SMTAttributes
+) where
+
+import           Data.Default
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+
+import           Kore.AST.Common
+import           Kore.AST.MetaOrObject
+import           Kore.AST.PureML
+import           Kore.ASTUtils.SmartPatterns
+import           Kore.ASTUtils.Substitution
+import           Kore.Attribute.Parser
+                 ( ParseAttributes (..) )
+import           Kore.Builtin.Hook
+import           Kore.IndexedModule.MetadataTools
+import qualified Kore.Predicate.Predicate as KorePredicate
+
+import           Data.Reflection
+import           Data.SBV
+
+import           GHC.IO.Unsafe
+
+data SMTAttributes
+  = SMTAttributes
+  { hook :: !Hook
+  } deriving(Eq, Show)
+
+instance Default SMTAttributes where
+    def = SMTAttributes def
+
+instance ParseAttributes SMTAttributes where
+    attributesParser = do
+        hook <- attributesParser
+        pure SMTAttributes {..} 
+
+provePredicate
+    :: Given (MetadataTools Object SMTAttributes)
+    => KorePredicate.CommonPredicate Object
+    -> Maybe Bool
+provePredicate p = provePattern $ KorePredicate.unwrapPredicate p
+
+-- | Returns `Just True` if the sentence is satisfied in all models,
+-- `Just False` if it has a counterexample,
+-- Nothing if timeout/undecidable.
+provePattern
+    :: Given (MetadataTools Object SMTAttributes)
+    => CommonPurePattern Object
+    -> Maybe Bool
+provePattern p =
+    case res of
+      Unsatisfiable _   -> Just True
```

Comment:
You're using a lot of strange indentations in this file and its tests: 2 spaces instead of 4, 1 space instead of 2, 0 spaces instead of 4 (e.g. in `prop4`). Indenting a `case` at the top level in a function two spaces and then the case branches another two spaces (as done in `getSortAttributes`) is not that bad (although it is not explicitly allowed by the coding style doc), but the others are kind of bad.

If you think it's useful I'll point out all the issues, but it would be easier if you'd prune them first this function has two (the `case` branch indentation and the `where` one).